File: ios/bitlabs/Sources/bitlabs/BitlabsPlugin.swift
import BLCustom
import Flutter
import UIKit

public class BitlabsPlugin: NSObject, FlutterPlugin {
    var offerwall: Offerwall?
    
    let channel: FlutterMethodChannel
    
    init(channel: FlutterMethodChannel) {
        self.channel = channel
        super.init()
    }
    
    
    public static func register(with registrar: FlutterPluginRegistrar) {
        let ch = FlutterMethodChannel(name: "bitlabs", binaryMessenger: registrar.messenger())
        let instance = BitlabsPlugin(channel: ch)
        registrar.addMethodCallDelegate(instance, channel: ch)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "init": initImpl(call, result)
        case "addTag": addTagImpl(call, result)
        case "setTags": setTagsImpl(call, result)
        case "launchOfferWall": launchOfferWallImpl(result)
        case "getSurveys": getSurveysImpl(result)
        case "checkSurveys": checkSurveysImpl(result)
        case "openOffer": openOfferImpl(call, result)
        case "openMagicReceiptsOffer": openMagicReceiptsOfferImpl(call, result)
        case "openMagicReceiptsMerchant": openMagicReceiptsMerchantImpl(call, result)
        case "requestTrackingAuthorization": requestTrackingAuthorizationImpl(result)
        default: result(FlutterMethodNotImplemented)
        }
    }

    private func initImpl(_ call: FlutterMethodCall, _ result: FlutterResult) {
        guard let arguments = call.arguments as? [String: Any],
              let token = arguments["token"] as? String,
              let uid = arguments["uid"] as? String
        else {
            result(FlutterError(code: "Error", message: "Invalid arguments", details: nil))
            return
        }
        
        offerwall = BitLabs.OFFERWALL.create(token: token, uid: uid)
        
        offerwall?.offerwallClosedHandler = { reward in
            self.channel.invokeMethod("onReward", arguments: ["reward": reward])
        }
        result("BitLabs initialized")
    }

    private func launchOfferWallImpl(_ result: FlutterResult) {
        guard let topViewController = UIApplication.shared.keyWindow?.rootViewController else {
            result(FlutterError(code: "Error", message: "Unable to access UIApplication's window", details: nil))
            return
        }

        offerwall?.launch(parent: topViewController)
        result("Offer wall launched")
    }

    private func addTagImpl(_ call: FlutterMethodCall, _ result:FlutterResult) {
        guard let arguments = call.arguments as? [String: Any],
              let key = arguments["key"] as? String,
              let value = arguments["value"] as? String
        else {
            result(FlutterError(code: "Error", message: "Invalid arguments", details: nil))
            return
        }

        offerwall?.tags[key] = value
    }

    private func setTagsImpl(_ call: FlutterMethodCall, _ result: FlutterResult) {
        guard let arguments = call.arguments as? [String: Any],
              let tags = arguments["tags"] as? [String: Any]
        else {
            result(FlutterError(code: "Error", message: "Invalid tags argument", details: nil))
            return
        }

        offerwall?.tags = tags
        result("Tags set")
    }

    private func getSurveysImpl(_ result: @escaping FlutterResult) {
        BitLabs.shared.getSurveys { response in
            switch response {
            case .failure(let error):
                result(FlutterError(code: "Error", message: "Get Surveys \(error)", details: nil))

            case .success(let surveys):
                if let data = try? JSONEncoder().encode(surveys),
                   let json = try? JSONSerialization.jsonObject(with: data, options: [])
                {
                    result(json)
                }

                result(FlutterError(code: "Error", message: "Couldn't Serialize the surveys", details: nil))
            }
        }
    }

    private func checkSurveysImpl(_ result: @escaping FlutterResult) {
        BitLabs.shared.checkSurveys { response in
            switch response {
            case .failure(let error):
                result(FlutterError(code: "Error", message: "Check Surveys \(error)", details: nil))

            case .success(let surveysExist):
                result(surveysExist)
            }
        }
    }

    private func openOfferImpl(_ call: FlutterMethodCall, _ result: FlutterResult) {
        guard let arguments = call.arguments as? [String: Any],
              let offerId = arguments["offerId"] as? String,
              let topViewController = UIApplication.shared.keyWindow?.rootViewController
        else {
            result(FlutterError(code: "Error", message: "Invalid arguments", details: nil))
            return
        }

        offerwall?.openOffer(withId: offerId, parent: topViewController)
        result("Offer opened")
    }

    private func openMagicReceiptsOfferImpl(_ call: FlutterMethodCall, _ result: FlutterResult) {
        guard let arguments = call.arguments as? [String: Any],
              let receiptId = arguments["offerId"] as? String,
              let topViewController = UIApplication.shared.keyWindow?.rootViewController
        else {
            result(FlutterError(code: "Error", message: "Invalid arguments", details: nil))
            return
        }

        offerwall?.openMagicReceiptsOffer(withId: receiptId, parent: topViewController)
        result("Magic Receipts Offer opened")
    }

    private func openMagicReceiptsMerchantImpl(_ call: FlutterMethodCall, _ result: FlutterResult) {
        guard let arguments = call.arguments as? [String: Any],
              let merchantId = arguments["merchantId"] as? String,
              let topViewController = UIApplication.shared.keyWindow?.rootViewController
        else {
            result(FlutterError(code: "Error", message: "Invalid arguments", details: nil))
            return
        }

        offerwall?.openMagicReceiptsMerchant(withId: merchantId, parent: topViewController)
        result("Magic Receipts Merchant opened")
    }

    private func requestTrackingAuthorizationImpl(_ result: @escaping FlutterResult) {
        offerwall?.requestTrackingAuthorization()
        result(true)
    }
}

File: android/src/main/kotlin/ai/bitlabs/bitlabs_flutter/BitlabsPlugin.kt
package ai.bitlabs.bitlabs_flutter

import android.app.Activity
import ai.bitlabs.sdk.BitLabs
import ai.bitlabs.sdk.offerwall.Offerwall
import ai.bitlabs.sdk.util.OnOfferwallClosedListener
import android.util.Log
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

/** BitlabsPlugin */
class BitlabsPlugin : FlutterPlugin, MethodCallHandler, ActivityAware {
    private lateinit var channel: MethodChannel
    private lateinit var offerwall: Offerwall

    private var activity: Activity? = null

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "bitlabs")
        channel.setMethodCallHandler(this)
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "init" -> initImpl(call, result)
            "setTags" -> setTagsImpl(call, result)
            "addTag" -> addTagImpl(call, result)
            "launchOfferWall" -> launchOfferWallImpl(result)
            "getSurveys" -> getSurveysImpl(result)
            "checkSurveys" -> checkSurveysImpl(result)
            "openOffer" -> openOfferImpl(call, result)
            "openMagicReceiptsOffer" -> openMagicReceiptsOfferImpl(call, result)
            "openMagicReceiptsMerchant" -> openMagicReceiptsMerchantImpl(call, result)
            else -> result.notImplemented()
        }
    }

    private fun initImpl(call: MethodCall, result: Result) = activity?.let {
        val token = call.argument<String>("token") ?: ""
        val uid = call.argument<String>("uid") ?: ""
        offerwall = BitLabs.OFFERWALL.create(token, uid)

        offerwall.onOfferwallClosedListener = OnOfferwallClosedListener { totalReward ->
            channel.invokeMethod("onReward", mapOf("reward" to totalReward))
        }

        BitLabs.API.init(token, uid)
        result.success("BitLabs initialised")
    } ?: result.error("Error", "Activity is null", null)

    private fun setTagsImpl(call: MethodCall, result: Result) {
        val tags = call.argument<Map<String, Any>>("tags") ?: run {
            result.error("Error", "Invalid tags argument", null)
            return
        }

        offerwall.tags.putAll(tags.toMutableMap())
        result.success("Tags set")
    }

    private fun addTagImpl(call: MethodCall, result: Result) {
        val key = call.argument<String>("key") ?: run {
            result.error("Error", "Invalid key argument", null)
            return
        }
        val value = call.argument<String>("value") ?: run {
            result.error("Error", "Invalid value argument", null)
            return
        }

        offerwall.tags[key] = value
        result.success("Tag added")
    }

    private fun launchOfferWallImpl(result: Result) = activity?.let {
        offerwall.launch(it)
        result.success("Offerwall launched")
    } ?: result.error("Error", "Activity is null", null)

    private fun getSurveysImpl(result: Result) = BitLabs.API.getSurveys(
        { surveys ->
            val surveysMap = surveys.map { survey -> survey.toMap() }
            result.success(surveysMap)
        },
        { exception -> result.error("Error", exception.message, null) }
    )

    private fun checkSurveysImpl(result: Result) = BitLabs.API.checkSurveys(
        { surveysExist -> result.success(surveysExist) },
        { exception -> result.error("Error", exception.message, null) }
    )

    private fun openOfferImpl(call: MethodCall, result: Result) = activity?.let {
        val offerId = call.argument<String>("offerId") ?: run {
            result.error("Error", "Invalid offerId argument", null)
            return
        }
        offerwall.openOffer(it, offerId)
        result.success("Offer opened")
    } ?: result.error("Error", "Activity is null", null)

    private fun openMagicReceiptsOfferImpl(call: MethodCall, result: Result) = activity?.let {
        val offerId = call.argument<String>("offerId") ?: run {
            result.error("Error", "Invalid offerId argument", null)
            return
        }
        offerwall.openMagicReceiptsOffer(it, offerId)
        result.success("Magic Receipts Offer opened")
    } ?: result.error("Error", "Activity is null", null)

    private fun openMagicReceiptsMerchantImpl(call: MethodCall, result: Result) = activity?.let {
        val merchantId = call.argument<String>("merchantId") ?: run {
            result.error("Error", "Invalid merchantId argument", null)
            return
        }
        offerwall.openMagicReceiptsMerchant(it, merchantId)
        result.success("Magic Receipts Merchant opened")
    } ?: result.error("Error", "Activity is null", null)

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }

    // ActivityAware methods
    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activity = binding.activity
    }

    override fun onDetachedFromActivityForConfigChanges() {
        activity = null
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        activity = binding.activity
    }

    override fun onDetachedFromActivity() {
        activity = null
    }
}
